src/Workspaces/Remote/ServiceHub/Shared/Extensions.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal static class Extensions
+    {
+        public static async Task InvokeAsync(
+            this JsonRpc rpc, string targetName, IEnumerable<object> arguments,
+            Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync, CancellationToken cancellationToken)
+        {
+            Task task = null;
+
+            try
+            {
+                using (var mergedCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
```

Comment:
where?  you're literally calling "CancellationTokenSource.CreateLinkedTokenSource(cancellationToken)".  This is a static method that should be given multiple cancellation tokens to.